server/src/main/java/org/elasticsearch/search/vectors/VectorSimilarityQueryBuilder.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.TransportVersion;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public class VectorSimilarityQueryBuilder extends AbstractQueryBuilder<VectorSimilarityQueryBuilder> {
+    public static final String NAME = "vector_similarity";
+    private static final ParseField FIELD = new ParseField("field");
+    private static final ParseField QUERY_VECTOR = new ParseField("query_vector");
+    private static final ParseField NUM_CANDIDATES = new ParseField("num_candidates");
```

Comment:
I think I got confused because I thought that `num_candidates` in the `knn` query was to control the visited limit of:
```
public abstract TopDocs searchNearestVectors(
      String field, byte[] target, int k, Bits acceptDocs, int visitedLimit) throws IOException;
```
But in fact this visitedLimited is controlled by the cost of the filter in our implementation so `num_candidate` in knn query really means how many candidates we return per shard to compute the global top k.
I am failing to understand why though. If the `knn` query returns 10k docs per shard and then it is reduced by the coordinator to a top k, why would returning more than topk docs per shard helps? I am sure I am missing something obvious so sorry in advance for the noob question.

Reply:
Ok I think I got it now, for knn the num_candidates is to allow more post filtering locally on the shard and what is sent back to the coordinating node is the final shard topK after applying the post filter. This global topK is really a pain to understand correctly for me but that's maybe just because I am slow ;).